lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -2850,6 +2833,23 @@ void ConformanceChecker::checkNonFinalClassWitness(ValueDecl *requirement,
   }
 }
 
+ResolveWitnessResult
+ConformanceChecker::resolveWitnessAsOpaque(ValueDecl *requirement) {
+  assert(!isa<AssociatedTypeDecl>(requirement) && "Use resolveTypeWitnessVia*");
```

Comment:
What's the reason this doesn't just take an `AssociatedTypeDecl`?

Reply:
The whole conformance-checking logic is split between type witnesses and non-type witnesses, and in practice it makes sense because they're resolved pretty differently. In this case, you can't actually make an associated type opaque, so the assertion makes sense.